CSRFFilter should have set the attribute csrf-token, but it is null after upgrade to 3.6.4

### Describe the bug

I upgraded from 3.6.3 to 3.6.4, and since then the CsrfTokenParameterProvider getToken() fails 'CSRFFilter should have set the attribute csrf-token, but it is null ' after the first load of the page, subsequent load or GET requests fails with the same error 

![image](https://github.com/quarkusio/quarkus/assets/33322056/057840c2-423c-4ca1-9ae0-5a453fa69090)

2023-12-23 19:46:30,719 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (vert.x-eventloop-thread-0) HTTP Request to /api/table failed, error id: ace42e7c-bf2e-495b-b24f-52711085a983-2: io.quarkus.qute.TemplateException
        at io.quarkus.qute.CompletedStage.get(CompletedStage.java:65)
        at io.quarkus.qute.MultiResultNode.process(MultiResultNode.java:20)
        at io.quarkus.qute.MultiResultNode.process(MultiResultNode.java:20)
        at io.quarkus.qute.MultiResultNode.process(MultiResultNode.java:20)
        at io.quarkus.qute.TemplateImpl$TemplateInstanceImpl.lambda$renderData$5(TemplateImpl.java:239)
        at io.quarkus.qute.CompletedStage.whenComplete(CompletedStage.java:285)
        at io.quarkus.qute.TemplateImpl$TemplateInstanceImpl.renderData(TemplateImpl.java:233)
        at io.quarkus.qute.TemplateImpl$TemplateInstanceImpl.renderAsyncNoTimeout(TemplateImpl.java:224)
        at io.smallrye.context.impl.wrappers.SlowContextualSupplier.get(SlowContextualSupplier.java:21)
        at io.smallrye.mutiny.operators.uni.builders.UniCreateFromCompletionStage.subscribe(UniCreateFromCompletionStage.java:24)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.operators.uni.UniFailOnTimeout.subscribe(UniFailOnTimeout.java:36)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.groups.UniSubscribe.withSubscriber(UniSubscribe.java:51)
        at io.smallrye.mutiny.groups.UniSubscribe.with(UniSubscribe.java:110)
        at io.smallrye.mutiny.groups.UniSubscribe.with(UniSubscribe.java:88)
        at org.jboss.resteasy.reactive.server.handlers.UniResponseHandler.handle(UniResponseHandler.java:19)
        at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:150)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
        at org.jboss.resteasy.reactive.server.handlers.RestInitialHandler.beginProcessing(RestInitialHandler.java:48)
        at org.jboss.resteasy.reactive.server.vertx.ResteasyReactiveVertxHandler.handle(ResteasyReactiveVertxHandler.java:23)
        at org.jboss.resteasy.reactive.server.vertx.ResteasyReactiveVertxHandler.handle(ResteasyReactiveVertxHandler.java:10)
        at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1286)
        at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:177)
        at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:144)
        at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:102)
        at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:88)
        at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1286)
        at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:140)
        at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:144)
        at io.vertx.ext.web.handler.impl.StaticHandlerImpl.lambda$sendStatic$1(StaticHandlerImpl.java:290)
        at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:141)
        at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:54)
        at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.IllegalStateException: CSRFFilter should have set the attribute csrf-token, but it is null
        at io.quarkus.csrf.reactive.runtime.CsrfTokenParameterProvider.getToken(CsrfTokenParameterProvider.java:42)
        at io.quarkus.csrf.reactive.runtime.CsrfTokenParameterProvider_ClientProxy.getToken(Unknown Source)
        at io.quarkus.csrf.reactive.runtime.CsrfTokenParameterProvider_ValueResolver.resolve(Unknown Source)
        at io.quarkus.qute.EvaluatorImpl.resolve(EvaluatorImpl.java:211)
        at io.quarkus.qute.EvaluatorImpl.resolveReference(EvaluatorImpl.java:131)
        at io.quarkus.qute.EvaluatorImpl.lambda$evaluate$0(EvaluatorImpl.java:78)
        at io.quarkus.qute.CompletedStage.thenCompose(CompletedStage.java:249)
        at io.quarkus.qute.EvaluatorImpl.evaluate(EvaluatorImpl.java:77)
        at io.quarkus.qute.ResolutionContextImpl$ChildResolutionContext.evaluate(ResolutionContextImpl.java:87)
        at io.quarkus.qute.ExpressionNode.resolve(ExpressionNode.java:36)
        at io.quarkus.qute.SectionNode$SectionResolutionContextImpl.execute(SectionNode.java:228)
        at io.quarkus.qute.SectionHelper$SectionResolutionContext.execute(SectionHelper.java:76)
        at io.quarkus.qute.LoopSectionHelper.nextElement(LoopSectionHelper.java:141)
        at io.quarkus.qute.LoopSectionHelper.lambda$resolve$0(LoopSectionHelper.java:58)
        at io.quarkus.qute.CompletedStage.thenCompose(CompletedStage.java:249)
        at io.quarkus.qute.LoopSectionHelper.resolve(LoopSectionHelper.java:47)
        at io.quarkus.qute.SectionNode.resolve(SectionNode.java:53)
        at io.quarkus.qute.SectionNode.resolve(SectionNode.java:58)
        at io.quarkus.qute.SectionNode$SectionResolutionContextImpl.execute(SectionNode.java:228)
        at io.quarkus.qute.SectionHelper$SectionResolutionContext.execute(SectionHelper.java:66)
        at io.quarkus.qute.Parser$1.resolve(Parser.java:1288)
        at io.quarkus.qute.SectionNode.resolve(SectionNode.java:53)
        at io.quarkus.qute.SectionNode.resolve(SectionNode.java:58)
        ... 35 more

![image](https://github.com/quarkusio/quarkus/assets/33322056/6ae401bf-0973-4508-a385-367b26d6cb28)


### Expected behavior

The CSRF token should be in context

### Actual behavior

The CSRF token is not in contett

### How to Reproduce?

clone https://github.com/yazalulloa/quarkus-csrf-signature-bug

Run quarkus project

Go to http://localhost:9898/some-page

### Output of `uname -a` or `ver`

Linux yaz-laptop 6.2.0-39-generic #40~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Nov 16 10:53:04 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "21" 2023-09-19 OpenJDK Runtime Environment GraalVM CE 21+35.1 (build 21+35-jvmci-23.1-b15) OpenJDK 64-Bit Server VM GraalVM CE 21+35.1 (build 21+35-jvmci-23.1-b15, mixed mode, sharing)

### Quarkus version or git rev

3.6.4

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

maven 3.9.3

### Additional information

_No response_

--- a/extensions/csrf-reactive/runtime/src/main/java/io/quarkus/csrf/reactive/runtime/CsrfReactiveConfig.java
+++ b/extensions/csrf-reactive/runtime/src/main/java/io/quarkus/csrf/reactive/runtime/CsrfReactiveConfig.java
@@ -34,7 +34,7 @@ public class CsrfReactiveConfig {
     /**
      * CSRF cookie max age.
      */
-    @ConfigItem(defaultValue = "10M")
+    @ConfigItem(defaultValue = "2H")
     public Duration cookieMaxAge;
 
     /**

--- a/extensions/csrf-reactive/runtime/src/main/java/io/quarkus/csrf/reactive/runtime/CsrfRequestResponseReactiveFilter.java
+++ b/extensions/csrf-reactive/runtime/src/main/java/io/quarkus/csrf/reactive/runtime/CsrfRequestResponseReactiveFilter.java
@@ -30,6 +30,7 @@ public class CsrfRequestResponseReactiveFilter {
      */
     private static final String CSRF_TOKEN_KEY = "csrf_token";
     private static final String CSRF_TOKEN_BYTES_KEY = "csrf_token_bytes";
+    private static final String NEW_COOKIE_REQUIRED = "true";
 
     /**
      * CSRF token verification status.
@@ -85,12 +86,14 @@ public class CsrfRequestResponseReactiveFilter {
 
         if (requestMethodIsSafe(requestContext)) {
             // safe HTTP method, tolerate the absence of a token
-            if (cookieToken == null && isCsrfTokenRequired(routing, config)) {
+            if (isCsrfTokenRequired(routing, config)) {
                 // Set the CSRF cookie with a randomly generated value
                 byte[] tokenBytes = new byte[config.tokenSize];
                 secureRandom.nextBytes(tokenBytes);
                 routing.put(CSRF_TOKEN_BYTES_KEY, tokenBytes);
                 routing.put(CSRF_TOKEN_KEY, Base64.getUrlEncoder().withoutPadding().encodeToString(tokenBytes));
+
+                routing.put(NEW_COOKIE_REQUIRED, true);
             }
         } else if (config.verifyToken) {
             // unsafe HTTP method, token is required
@@ -129,7 +132,6 @@ public class CsrfRequestResponseReactiveFilter {
             String csrfTokenFormParam = (String) rrContext.getFormParameter(config.formFieldName, true, false);
             LOG.debugf("CSRF token found in the form parameter");
             verifyCsrfToken(requestContext, routing, config, cookieToken, csrfTokenFormParam);
-            return;
 
         } else if (cookieToken == null) {
             LOG.debug("CSRF token is not found");
@@ -159,6 +161,8 @@ public class CsrfRequestResponseReactiveFilter {
             } else {
                 routing.put(CSRF_TOKEN_KEY, csrfToken);
                 routing.put(CSRF_TOKEN_VERIFIED, true);
+                // reset the cookie
+                routing.put(NEW_COOKIE_REQUIRED, true);
                 return;
             }
         }
@@ -195,9 +199,9 @@ public class CsrfRequestResponseReactiveFilter {
     @ServerResponseFilter
     public void filter(ContainerRequestContext requestContext,
             ContainerResponseContext responseContext, RoutingContext routing) {
-        final CsrfReactiveConfig config = configInstance.get();
-        if (requestContext.getMethod().equals("GET") && isCsrfTokenRequired(routing, config)
-                && getCookieToken(routing, config) == null) {
+        if (routing.get(NEW_COOKIE_REQUIRED) != null) {
+
+            final CsrfReactiveConfig config = configInstance.get();
 
             String cookieValue = null;
             if (config.tokenSignatureKey.isPresent()) {
@@ -230,7 +234,7 @@ public class CsrfRequestResponseReactiveFilter {
      *
      * @return An Optional containing the token, or an empty Optional if the token cookie is not present or is invalid
      */
-    private String getCookieToken(RoutingContext routing, CsrfReactiveConfig config) {
+    private static String getCookieToken(RoutingContext routing, CsrfReactiveConfig config) {
         Cookie cookie = routing.getCookie(config.cookieName);
 
         if (cookie == null) {
@@ -241,14 +245,14 @@ public class CsrfRequestResponseReactiveFilter {
         return cookie.getValue();
     }
 
-    private boolean isCsrfTokenRequired(RoutingContext routing, CsrfReactiveConfig config) {
+    private static boolean isCsrfTokenRequired(RoutingContext routing, CsrfReactiveConfig config) {
         return config.createTokenPath
                 .map(value -> value.contains(routing.normalizedPath())).orElse(true);
     }
 
-    private void createCookie(String csrfToken, RoutingContext routing, CsrfReactiveConfig config) {
+    private static void createCookie(String cookieTokenValue, RoutingContext routing, CsrfReactiveConfig config) {
 
-        ServerCookie cookie = new CookieImpl(config.cookieName, csrfToken);
+        ServerCookie cookie = new CookieImpl(config.cookieName, cookieTokenValue);
         cookie.setHttpOnly(config.cookieHttpOnly);
         cookie.setSecure(config.cookieForceSecure || routing.request().isSSL());
         cookie.setMaxAge(config.cookieMaxAge.toSeconds());

--- a/integration-tests/csrf-reactive/src/test/java/io/quarkus/it/csrf/CsrfReactiveTest.java
+++ b/integration-tests/csrf-reactive/src/test/java/io/quarkus/it/csrf/CsrfReactiveTest.java
@@ -41,6 +41,8 @@ public class CsrfReactiveTest {
         try (final WebClient webClient = createWebClient()) {
             webClient.addRequestHeader("Authorization", basicAuth("alice", "alice"));
             HtmlPage htmlPage = webClient.getPage("http://localhost:8081/service/csrfTokenForm");
+            htmlPage = webClient.getPage("http://localhost:8081/service/csrfTokenForm");
+            assertNotNull(htmlPage.getWebResponse().getResponseHeaderValue("Set-Cookie"));
 
             assertEquals("CSRF Token Form Test", htmlPage.getTitleText());
 
@@ -51,11 +53,19 @@ public class CsrfReactiveTest {
             assertNotNull(webClient.getCookieManager().getCookie("csrftoken"));
 
             TextPage textPage = loginForm.getInputByName("submit").click();
-
+            assertNotNull(htmlPage.getWebResponse().getResponseHeaderValue("Set-Cookie"));
             assertEquals("alice:true:tokenHeaderIsSet=false", textPage.getContent());
 
+            // This request which returns String is not CSRF protected
             textPage = webClient.getPage("http://localhost:8081/service/hello");
             assertEquals("hello", textPage.getContent());
+            // therefore no Set-Cookie header is expected
+            assertNull(textPage.getWebResponse().getResponseHeaderValue("Set-Cookie"));
+
+            // Repeat a form submission
+            textPage = loginForm.getInputByName("submit").click();
+            assertNotNull(htmlPage.getWebResponse().getResponseHeaderValue("Set-Cookie"));
+            assertEquals("alice:true:tokenHeaderIsSet=false", textPage.getContent());
 
             webClient.getCookieManager().clearCookies();
         }

